### PR TITLE
#0: Simplify erisc entry/exit

### PR DESCRIPTION
**Pull request by nathan-TT**
### Ticket
NA

### Problem description
In investigating execute-in-place code gen I noticed the entry sequence here was  a home-brewed setjmp/longjmp. I determined why sj-lj wasn't used, so commented on that.  But I also noticed we ended up with two instances of longjmp -- the one in the regular exit path and the separate callable one. Besides being code bloat, that also means we have to get two non-adjacent things consistent, which is asking for trouble.

### What's changed
1) Do not use `#ifdef __cplusplus` inside a C++ source file -- just use C++.  Convert `extern "C" { ... }` to plain `extern "C" $decl`.
2) No need to define `CONTEXT_SIZE`, it's only used inside the `ApplicationHandler` function, and highly dependent on that implementation.
3) Reimplement `ApplicationHandler`.  Separate out the various asm fragments, use the `s$n` callee-save register names. Embed longjmp directly in the exit path.
4) Delete the separate `erisc-early-exit.S` implementation.

I'm not sure the `erisc_early_exit` entry point actually needs a separate parameter informing it of the save slot -- we only write the save slot in one place.  I've not changed that though.

### Checklist
- [YES ] Post commit CI passes
- [YES ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ YES] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by nathan-TT** [comment]
And alternative to burying `erisc_early_exit` inside `ApplicationHandler` would be a separate naked function and a tail call.

**Review by pgkeller — APPROVED** [review]
Please have @ubcheema review as well

File: tt_metal/hw/firmware/src/erisc.cc [hunk 1]
```diff
@@ -92,49 +81,80 @@ void __attribute__((section("erisc_l1_code.1"), noinline)) Application(void) {
...
+  // Place the erisc_early_exit inside this function! avoids a tail call.
+  // void erisc_early_exit(uint32_t save_slot)
+  __asm__ volatile(".global erisc_early_exit\n\t"
+	  ".type erisc_early_exit,@function\n\t"
+	  "erisc_early_exit:\n\t");
+
+  { // erisc_early_exit pseudo-scope
+
+    // Restore sp from the save slot.
+    __asm__ volatile("lw sp, 0(a0)\n\t");
```
[the remark this answers is not in the document]
Comment (nathan-TT) [reply tt_metal/hw/firmware/src/erisc.cc:136]: a0 is the incoming argument register, that's the address holding the SP. In the non exceptional path that's set on line 124.  in the exceptional path, the caller will have set it.
What am I missing?

[the remark this answers is not in the document]
Comment (ubcheema) [reply tt_metal/hw/firmware/src/erisc.cc:136]: I see, so in exception case, we short circuit to line 136. From somewhere in the Application( ) downstream code?

**Review by ubcheema — APPROVED** [review]

File: tt_metal/hw/firmware/src/erisc.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathan-TT) [reply tt_metal/hw/firmware/src/erisc.cc:136]: correct. I don't know why the API is such that the exceptional path has to pass in the slot address -- it's not like there's a choice.
But your confusion suggests we should have two naked function defs here, and eat the cost of a tail call insn?

[the remark this answers is not in the document]
Comment (ubcheema) [reply tt_metal/hw/firmware/src/erisc.cc:136]: I don't think that's necessary.
I was confused about whether line 133 was a standalone function or not.
Looks like its inlined into ApplicationHandler( ). And line 129 is a hook to short circuit into it.

**Pull request merged**